vcd/provider.go

```diff
@@ -108,6 +109,33 @@ func Provider() terraform.ResourceProvider {
 	}
 }
 
+// This is a global MutexKV for all resources
+var vcdMutexKV = mutexkv.NewMutexKV()
+
+func lockVapp(d *schema.ResourceData) {
```

Comment:
Good question, indeed! Also, how do we handle different locking levels with a single global mutex map? There are the very least three of them:
1. vApp operations - on Org name (?), VDC name and vApp name
2. Edge Gateway operations - on Org name and EG name
3. Org VDC creation/destroy/update - on Org name alone? Or is it possible to create Org VDCs in parallel and no locking is needed at all?

Reply:
1. yes, locking on specific VAPP
2. yes we could for Edge gtw operations, but not creation, and yet no other operation at the moment. VDC networks create/delete locks on parent edge gtw.
3. Org VDC - dunno, haven't test or anyone complain about that.

Reply:
P.s. to be on safe side, will add org name to `key`

Reply:
> Please also add an antry in CHANGELOG.md for this feature.

Added

Reply:
Thanks. One nuance: I suggest having "org" first, "vdc" second and object name last in accordance with the hierarchy in vCD. It would be more intuitive.

Reply:
Changed

Reply:
regarding `3` point - I made check and it is possible to create org VDC in parallel.

Reply:
Regarding the logging. Need others to share their thoughts, but while debugging such messages can be confusing. Here's why:


`[TRACE] Locked vapp with key my-org|my-vdc||my-vapp`
You get this message, but actually the lock may still not be acquired, just stuck at waiting for it to be acquired. IMHO this can be very confusing when debugging logs.

I would propose something like.
```
key := fmt.Sprintf("%s|%s|%s", d.Get("org").(string), d.Get("vdc").(string), d.Get("name").(string))
log.Printf("[TRACE] Acquiring lock for vapp with key %s.", key)
vcdMutexKV.Lock(key)
log.Printf("[TRACE] Locked vapp with key %s.", key)
```

Reply:
One more thing, we discussed with Dainius that there's a risk to collide if vApp name and EG name are the same. This is realistic - for example a name "Test". Hence, we're suggesting to add type to the string:
`org:myOrg|vdc:myVdc|edge:myTest`
`org:myOrg|vdc:myVdc|vapp:myTest`

Reply:
Changed

Reply:
Nit: please capitalise "vApp" in the logs as well. It's just so much easier to scan the text when it's everywhere a "vApp" instead of "vapp".

Reply:
I was too early to complain about log messages... Seems like the `mutexkv` library logs itself. Here's the output. You will see the `DEBUG` level messages are coming from library itself while `TRACE` ones are coming from @vbauzysvmware  code:
https://github.com/hashicorp/terraform/blob/master/helper/mutexkv/mutexkv.go#L21
```
2019-06-04T11:08:14.283+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:14 [TRACE] Acquiring lock for vapp with key org:my-org|vdc:my-org-vdc|vapp:TestAccVcdVAppVapp-other.
2019-06-04T11:08:14.283+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:14 [DEBUG] Locking "org:my-org|vdc:my-org-vdc|vapp:TestAccVcdVAppVapp-other"
2019-06-04T11:08:14.283+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:14 [DEBUG] Locked "org:my-org|vdc:my-org-vdc|vapp:TestAccVcdVAppVapp-other"
2019-06-04T11:08:14.283+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:14 [TRACE] Locked vapp with key org:my-org|vdc:my-org-vdc|vapp:TestAccVcdVAppVapp-other.
2019-06-04T11:08:17.811+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:17 [TRACE] Unlocked vapp with key org:my-org|vdc:my-org-vdc|vapp:TestAccVcdVAppVapp-other.
2019-06-04T11:08:17.811+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:17 [DEBUG] Unlocking "org:my-org|vdc:my-org-vdc|vapp:TestAccVcdVAppVapp-other"
2019-06-04T11:08:17.811+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:17 [DEBUG] Unlocked "org:my-org|vdc:my-org-vdc|vapp:TestAccVcdVAppVapp-other"
2019-06-04T11:08:17.824+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:17 [TRACE] Acquiring lock for edge gtw with key org:my-org|vdc:my-org-vdc|edge:my-edge-gw.
2019-06-04T11:08:17.824+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:17 [DEBUG] Locking "org:my-org|vdc:my-org-vdc|edge:my-edge-gw"
2019-06-04T11:08:17.824+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:17 [DEBUG] Locked "org:my-org|vdc:my-org-vdc|edge:my-edge-gw"
2019-06-04T11:08:17.824+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:17 [TRACE] Locked parent edge gtw with key org:my-org|vdc:my-org-vdc|edge:my-edge-gw.
2019-06-04T11:08:18.317+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:18 [TRACE] Unlocked parent edge gtw with key org:my-org|vdc:my-org-vdc|edge:my-edge-gw.
2019-06-04T11:08:18.317+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:18 [DEBUG] Unlocking "org:my-org|vdc:my-org-vdc|edge:my-edge-gw"
2019-06-04T11:08:18.317+0300 [DEBUG] plugin.terraform-provider-vcd_v2.3.0: 2019/06/04 11:08:18 [DEBUG] Unlocked "org:my-org|vdc:my-org-vdc|edge:my-edge-gw"
```

Probably logging from two places is overkill. What do others think?

Reply:
Ouch! Yes, now it's overkill.

BTW, it looks clear what's being locked with the new string format with types!

Reply:
Removed logs